src/sources/nginx_metrics/mod.rs

```diff
@@ -0,0 +1,316 @@
+        }
+
+        let mut out = out
+            .sink_map_err(|error| error!(message = "Error sending mongodb metrics.", %error))
+            .sink_compat();
+
+        let duration = time::Duration::from_secs(self.scrape_interval_secs);
+        Ok(Box::pin(async move {
+            let mut interval = time::interval(duration).take_until(shutdown);
+            while interval.next().await.is_some() {
+                let start = Instant::now();
+                let metrics = join_all(sources.iter().map(|nginx| nginx.collect())).await;
+                emit!(NginxMetricsCollectCompleted {
+                    start,
+                    end: Instant::now()
+                });
+
+                let mut stream = stream::iter(metrics).flatten().map(Event::Metric).map(Ok);
+                out.send_all(&mut stream).await?;
+            }
+
+            Ok(())
+        }))
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Metric
+    }
+
+    fn source_type(&self) -> &'static str {
+        "nginx_metrics"
+    }
+}
+
+#[derive(Debug)]
+struct NginxMetrics {
+    http_client: HttpClient,
+    endpoint: String,
+    auth: Option<Auth>,
+    namespace: Option<String>,
+    tags: BTreeMap<String, String>,
+}
+
+impl NginxMetrics {
+    fn new(
+        http_client: HttpClient,
+        endpoint: String,
+        auth: Option<Auth>,
+        namespace: Option<String>,
+    ) -> crate::Result<Self> {
+        let mut tags = BTreeMap::new();
+        tags.insert("endpoint".into(), endpoint.clone());
+        tags.insert("host".into(), Self::get_endpoint_host(&endpoint)?);
+
+        Ok(Self {
+            http_client,
+            endpoint,
+            auth,
+            namespace,
+            tags,
+        })
+    }
+
+    fn get_endpoint_host(endpoint: &str) -> crate::Result<String> {
+        let uri: Uri = endpoint.parse().context(HostInvalidUri)?;
+        Ok(match (uri.host().unwrap_or(""), uri.port()) {
+            (host, None) => format!("{}", host),
+            (host, Some(port)) => format!("{}:{}", host, port),
+        })
+    }
+
+    async fn collect(&self) -> stream::BoxStream<'static, Metric> {
+        let (up_value, metrics) = match self.collect_metrics().await {
+            Ok(metrics) => (1.0, metrics),
+            Err(()) => (0.0, vec![]),
+        };
+
+        stream::once(ready(self.create_metric("up", gauge!(up_value))))
+            .chain(stream::iter(metrics))
+            .boxed()
+    }
+
+    async fn collect_metrics(&self) -> Result<Vec<Metric>, ()> {
+        let response = self.get_nginx_response().await.map_err(|error| {
+            emit!(NginxMetricsRequestError {
+                error,
+                endpoint: &self.endpoint,
+            })
+        })?;
+
+        let status = match String::from_utf8_lossy(&response) {
+            Cow::Borrowed(data) => NginxStubStatus::try_from(data),
+            Cow::Owned(data) => NginxStubStatus::try_from(data.as_str()),
+        }
```

Comment:
I believe this can be simply:
```suggestion
        let status = NginxStubStatus::try_from(&String::from_utf8_lossy(&response))
```
(`Cow` implements `AsRef` for `str`)

Reply:
Yep, I know that `Cow` impl `AsRef`, but when I tried `&String::from_utf8_lossy(&response)` compiler produced:
```
error[E0277]: the trait bound `NginxStubStatus: From<&Cow<'_, str>>` is not satisfied
   --> src/sources/nginx_metrics/mod.rs:199:13
    |
199 |             NginxStubStatus::try_from(&String::from_utf8_lossy(&response)).map_err(|error| {
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `From<&Cow<'_, str>>` is not implemented for `NginxStubStatus`
    |
    = note: required because of the requirements on the impl of `Into<NginxStubStatus>` for `&Cow<'_, str>`
    = note: required because of the requirements on the impl of `std::convert::TryFrom<&Cow<'_, str>>` for `NginxStubStatus`

error[E0308]: mismatched types
   --> src/sources/nginx_metrics/mod.rs:201:21
    |
201 |                     error,
    |                     ^^^^^ expected enum `sources::nginx_metrics::parser::ParseError`, found enum `Infallible`

error[E0277]: the trait bound `sources::nginx_metrics::parser::NginxStubStatus: From<&Cow<'_, str>>` is not satisfied
   --> src/sources/nginx_metrics/mod.rs:199:13
    |
199 |             NginxStubStatus::try_from(&String::from_utf8_lossy(&response)).map_err(|error| {
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `From<&Cow<'_, str>>` is not implemented for `sources::nginx_metrics::parser::NginxStubStatus`
    |
    = note: required because of the requirements on the impl of `Into<sources::nginx_metrics::parser::NginxStubStatus>` for `&Cow<'_, str>`
    = note: required because of the requirements on the impl of `std::convert::TryFrom<&Cow<'_, str>>` for `sources::nginx_metrics::parser::NginxStubStatus`

error[E0308]: mismatched types
   --> src/sources/nginx_metrics/mod.rs:201:21
    |
201 |                     error,
    |                     ^^^^^ expected enum `sources::nginx_metrics::parser::ParseError`, found enum `Infallible`
```

Tried `.as_ref()` directly and all works: `NginxStubStatus::try_from(String::from_utf8_lossy(&response).as_ref())`. Will push fix.